cartridge/roles/expirationd.lua

```diff
@@ -0,0 +1,209 @@
+        end
+        local ok, res = get_param("options." .. opt, val, opts_map[opt])
+        if not ok then
+            error(res)
+        end
+        opts[opt] = res
+    end
+
+    return opts
+end
+
+local function get_task_config(task_conf)
+    local conf = table.deepcopy(task_conf)
+    local params_map = {
+        space_id = {required = true, types = {"n", "s"}},
+        is_expired = {required = true, types = {"f"}},
+        is_master_only = {required = false, types = {"b"}},
+        options = {required = false, types = {"t"}},
+    }
+    for k, _ in pairs(conf) do
+        if type(k) ~= "string" then
+            error(role_name .. ": param must be a string")
+        end
+        if params_map[k] == nil then
+            error(role_name .. ": unsupported param " .. k)
+        end
+    end
+
+    for param, opts in pairs(params_map) do
+        if opts.required and conf[param] == nil then
+            error(role_name .. ": " .. param .. " is required")
+        end
+        if conf[param] ~= nil then
+            local ok, res = get_param(param, conf[param], opts.types)
+            if not ok then
+                error(res)
+            end
+            conf[param] = res
+        end
+    end
+
+    conf.options = get_task_options(conf.options)
+    return conf
+end
+
+local function init()
+
+end
+
+local function validate_config(conf_new)
+    local conf = table.deepcopy(conf_new[role_name] or {})
+
+    for task_name, task_conf in pairs(conf) do
+        local ok, res = get_param("task name", task_name, {"s"})
+        if not ok then
+            error(res)
+        end
+        get_task_config(task_conf)
+    end
+
+    return true
+end
+
+local function apply_config(conf_new, opts)
+    local conf = table.deepcopy(conf_new[role_name] or {})
+
+    -- finishes tasks from an old configuration
+    for i=#started,1,-1 do
+        local task_name = started[i]
+        local ok, _ = pcall(expirationd.task, task_name)
+        if ok then
+            if conf[task_name] then
+                expirationd.task(task_name):stop()
+            else
+                expirationd.task(task_name):kill()
+            end
+        end
+        table.remove(started, i)
+    end
+
+    for task_name, task_conf in pairs(conf) do
+        task_conf = get_task_config(task_conf)
+
+        local skip = task_conf.is_master_only and not opts.is_master
+        if not skip then
+            local task = expirationd.start(task_name, task_conf.space_id,
+                                           task_conf.is_expired,
+                                           task_conf.options)
+            if task == nil then
+                error(role_name .. ": unable to start task " .. task_name)
+            end
+            table.insert(started, task_name)
+        end
+    end
+end
+
+local function stop()
+    for _, task_name in pairs(expirationd.tasks()) do
+        local task = expirationd.task(task_name)
+        task:stop()
```

Comment:
expirationd.task may raise an error and seems that it's not a good idea to raise an error on stop.

Reply:
We iterate through the list of tasks from `expirationd`. It should not return an error from expirationd.task() if there are no bugs in an internal logic of expirationd.

I think that's why we don't need to check for errors here. Am I wrong?

Reply:
But why are u using pcall at line 168?

Reply:
This list of tasks was not received from expiration directly. It is a list of tasks that we started from a configuration file.

Nothing prevents an user from killing a task that we started. In this case expirationd.task(task_name) may return an error. So we need to handle it:

https://github.com/tarantool/expirationd/blob/ca28829ad5931e4f5eabcb67801602d4ab7cbdad/cartridge/roles/expirationd.lua#L166-L169

A test for the code:

https://github.com/tarantool/expirationd/blob/ca28829ad5931e4f5eabcb67801602d4ab7cbdad/test/unit/role_test.lua#L645

https://github.com/tarantool/expirationd/blob/ca28829ad5931e4f5eabcb67801602d4ab7cbdad/test/unit/role_test.lua#L617-L649